### PR TITLE
Type hint fix

### PR DESCRIPTION
**Pull request by Kieran-Fishwick-TfN**
Describe Changes
---
updated multi tld API

Task Checklist
----
- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Is the documentation building correctly with a clean version number?
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `setup.cfg`

### CODE REVIEW

File: src/caf/distribute/gravity_model/multi_area.py [hunk 1]
```diff
@@ -243,35 +372,67 @@ def _create_seed_matrix(self, cost_distributions, cost_args, params_len):
...
+                )
+            if min_cost < min_binning:
+                LOG.warning(
+                    f"the min cost in the cost matrix for"
+                    f" category {dist.name}, was {min_cost}, "
+                    "whereas the lowest bin edge in cost"
+                    f" distribution was {min_binning}, "
+                    "you will not be fitting to trips"
+                    " with a cost less than the binning"
+                )
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:419]: Use the warnings module instead of logging because it gives the user more control over the warnings, they are captured by the logging from toolkit by default anyway so will still show up in any log files.

In future we might want to define custom GM warnings to make it even easier for the user to handle them in specific ways if needed.
```suggestion
            if max_cost > max_binning:
                warnings.warn(
                    f"the maximum cost in the cost matrix for"
                    f" category {dist.name}, was {max_cost}, "
                    "whereas the highest bin edge in cost"
                    f" distribution was {max_binning}, "
                    "you will not be fitting to trips"
                    " with a cost greater than the binning"
                )
            if min_cost < min_binning:
                warnings.warn(
                    f"the min cost in the cost matrix for"
                    f" category {dist.name}, was {min_cost}, "
                    "whereas the lowest bin edge in cost"
                    f" distribution was {min_binning}, "
                    "you will not be fitting to trips"
                    " with a cost less than the binning"
                )
```

File: src/caf/distribute/gravity_model/multi_area.py [hunk 2]
```diff
@@ -243,35 +372,67 @@ def _create_seed_matrix(self, cost_distributions, cost_args, params_len):
...
         default_retry: bool = True,
         verbose: int = 0,
         **kwargs,
-    ) -> dict[str, GravityModelCalibrateResults]:
-        params_len = len(self.dists[0].function_params)
+    ) -> dict[str | int, GravityModelCalibrateResults]:
+        params_len = len(distributions[0].function_params)
         ordered_init_params = []
-        for dist in self.dists:
+        # TODO validate cost distributions
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:390]: I think it would be cleaner if the validation of cost distributions was moved to another private method to keep calibrate simpler.

File: src/caf/distribute/gravity_model/multi_area.py [hunk 3]
```diff
@@ -427,21 +530,24 @@ def calibrate(
         `caf.distribute.furness.doubly_constrained_furness()`
         `scipy.optimize.least_squares()`
         """
-        for dist in self.dists:
+        for dist in distributions:
             self.cost_function.validate_params(dist.function_params)
         self._validate_running_log(running_log_path)
         self._initialise_internal_params()
         return self._calibrate(  # type: ignore
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:537]: Because `self._calibrate` is only called here and this method doesn't contain much else other than a call to `_calibrate` I think it would probably be simpler to move all the code from `_calibrate` to here and remove the separate method.

I made some comments to clean-up the `_calibrate` method which I still think would be helpful.

File: src/caf/distribute/gravity_model/multi_area.py [hunk 4]
```diff
@@ -338,87 +499,29 @@ def _calibrate(
...
-            - 0 (default) : work silently.
-            - 1 : display a termination report.
-            - 2 : display progress during iterations (not supported by ‘lm’ method).
-
-        kwargs:
-            Additional arguments passed to self.gravity_furness.
-            Empty by default. The calling signature is:
-            `self.gravity_furness(seed_matrix, **kwargs)`
-
+        distributions: MultiCostDistribution,
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:518]: Docstring should explain all the parameters here because there is nowhere else for the user to see that information, I think the parameters will be simpler if the GMParams dataclass mentioned above is created because the details of each of those parameters would be documented there instead of here.

File: src/caf/distribute/gravity_model/multi_area.py [hunk 5]
```diff
@@ -427,21 +530,24 @@ def calibrate(
...
+            running_log_path,
             *args,
-            running_log_path=running_log_path,
             **kwargs,
         )
 
     def _jacobian_function(
         self,
         init_params: list[float],
         cost_distributions: list[MultiCostDistribution],
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:547]: This type annotation should is now just the class not a list
```suggestion
        cost_distributions: MultiCostDistribution,
```

File: src/caf/distribute/gravity_model/multi_area.py [hunk 6]
```diff
@@ -427,21 +530,24 @@ def calibrate(
...
+        furness_tol: int,
         diff_step: float,
+        furness_jac: bool,
         running_log_path,
         params_len,
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:552]: Type annotations for these, I think path and int?
```suggestion
        running_log_path: Path,
        params_len: int,
```

File: src/caf/distribute/gravity_model/multi_area.py [hunk 7]
```diff
@@ -1,12 +1,14 @@
...
 # Built-Ins
+from __future__ import annotations
+
 import functools
 import logging
 import os
 from dataclasses import dataclass
 from pathlib import Path
-from typing import Any, Optional
+from typing import Any, Optional, Iterator
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:11]: For future reference its better to import modules directly rather than specific functions / classes as it keeps the namespace cleaner and also avoids issues in sphinx docs e.g. `import pathlib` instead of `from pathlib import Path`

Typing is the one exception as the convention for that module is to import the classes directly, to avoid sphinx docs picking these up they can be imported with private aliases e.g. `from typing import Any as _Any`

File: src/caf/distribute/gravity_model/multi_area.py [hunk 8]
```diff
@@ -504,7 +610,14 @@ def _jacobian_function(
...
 
     def _gravity_function(
-        self, init_params, cost_distributions, running_log_path, params_len, diff_step=0
+        self,
+        init_params,
+        cost_distributions,
+        furness_tol,
+        running_log_path,
+        params_len,
+        diff_step=0,
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:619]: Type annotations

File: src/caf/distribute/gravity_model/multi_area.py [hunk 9]
```diff
@@ -563,29 +676,35 @@ def _gravity_function(
         return achieved_residuals
 
     # pylint:enable=too-many-locals
-    def run(self):
+    def run(
+        self,
+        distributions: list[MultiCostDistribution],
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:681]: This shouldn't be a list anymore
```suggestion
        distributions: MultiCostDistribution,
```

File: src/caf/distribute/gravity_model/multi_area.py [hunk 10]
```diff
@@ -563,29 +676,35 @@ def _gravity_function(
...
+        running_log_path: Path,
+        furness_tol=DEFAULT_FURNESS_TOL,
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:683]: Missing type annotation
```suggestion
        furness_tol: float = DEFAULT_FURNESS_TOL,
```

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
I think the methodology looks good, I've mainly made comments around cleaning up the code and adding documentation.

Some more general things that need updating are:
- Running isort (specifically on multi_area)
- Fixing pylint issues, many of these are covered by comments for tidying up.
- A bunch of tests are failing due to removing the params argument from `MultiAreaGravityModelCalibrator` these can just be updated to reflect the API change
- One test is failing due to init params outside the new boundaries for tanner, the init parameter for the test is  4.7... and -1.6..., potentially just bringing those init params within the limits will solve the issue with the test (tests\gravity_model\data\small_and_simple\tanner\calibrate\init_params.json)

There's one additional comment on the code which I couldn't make because you haven't edited that line.
Line 726 in multi_area.py: `multi_area.gravity_model` function has the wrong type annotation for `cost_distributions` should be `MultiCostDistribution` not `list[MultiCostDistribution]`

File: src/caf/distribute/gravity_model/multi_area.py [hunk 11]
```diff
@@ -156,39 +283,41 @@ def __init__(
...
+            if np.isinf(data).any():
+                raise ValueError(f"There are Infs in {name}")
+
+            num_zeros = (data == 0).sum()  # casting bool as 1, 0
+
+            LOG.info(f"There are {num_zeros} 0s in {name} ({(num_zeros/data.size)*100} %)")
+
+        zero_in_both = np.stack([row_targets == 0, col_targets == 0], axis=1).all(axis=1).sum()
+
+        LOG.info(f"There are {zero_in_both} zones with both 0 row and column targets.")
```
[the remark this answers is not in the document]
Comment (Kieran-Fishwick-TfN) [reply src/caf/distribute/gravity_model/multi_area.py:313]: I have made the change, but I want you know know I hate this format

File: src/caf/distribute/gravity_model/multi_area.py [hunk 12]
```diff
@@ -243,35 +372,67 @@ def _create_seed_matrix(self, cost_distributions, cost_args, params_len):
...
+        distributions: MultiCostDistribution,
+        running_log_path: Path,
+        furness_jac: bool = False,
         diff_step: float = 1e-8,
         ftol: float = 1e-4,
         xtol: float = 1e-4,
+        furness_tol=DEFAULT_FURNESS_TOL,
         grav_max_iters: int = 100,
         failure_tol: float = 0,
         default_retry: bool = True,
```
[the remark this answers is not in the document]
Comment (Kieran-Fishwick-TfN) [reply src/caf/distribute/gravity_model/multi_area.py:384]: I am concerned if we do this we would essentially have 3 input data classes nested in each other, which could get very confusing for those not familar with the code base

File: src/caf/distribute/gravity_model/multi_area.py (hunk 3 above)
Comment (Kieran-Fishwick-TfN) [reply src/caf/distribute/gravity_model/multi_area.py:649]: I think this is likely to change the API significantly, is this more a to do and new major verision number since it will be incompatible with previous versions

Comment (Kieran-Fishwick-TfN) [reply src/caf/distribute/gravity_model/multi_area.py:649]: wasn't the idea with this wrapper to hide the more complicated args to avoid confusion?

File: src/caf/distribute/gravity_model/multi_area.py [hunk 13]
```diff
@@ -92,6 +98,83 @@ class MultiDistInput(BaseConfig):
...
+        tld_avg_col: str = "av_distance",
+        tld_trips_col: str = "trips",
+        lookup_cat_col: str = "category",
+        lookup_zone_col: str = "zone_id",
+    ):
+
+        distributions: list[MGMCostDistribution] = []
+
+        for category in cat_zone_correspondence[lookup_cat_col].unique():
+
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply src/caf/distribute/gravity_model/multi_area.py]: @Kieran-Fishwick-TfN any reason this was missed?

File: .gitignore [hunk 14]
```diff
@@ -131,3 +131,6 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+
+
+settings.json
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply .gitignore]: @Kieran-Fishwick-TfN why was this missed?

File: src/caf/distribute/gravity_model/multi_area.py [hunk 15]
```diff
@@ -156,61 +501,51 @@ def __init__(
         col_targets: np.ndarray,
         cost_matrix: np.ndarray,
         cost_function: cost_functions.CostFunction,
-        params: Optional[MultiDistInput],
+        # TODO(kf) move these parameters as inputs of calibrate and run
```
Comment (MattBuckley-TfN) [inline src/caf/distribute/gravity_model/multi_area.py:504]: I think this has been done. hasn't it?

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Couple of previous comments haven't been addressed and one new one about a TODO that I think has been completed

**Review by MattBuckley-TfN — APPROVED** [review]
